packages/core/src/types.ts

```diff
@@ -380,3 +379,164 @@ export type BackgroundImageOptions = {
+type DesignTokensPayload = DesignTokensDefinition;
+type RegisteredBreakpointsPayload = { breakpoints: Breakpoint[] };
+type MouseMovePayload = { clientX: number; clientY: number };
+type NewHoveredElementPayload = { nodeId?: string };
+type ComponentSelectedPayload = {
+  nodeId: string;
+  assembly?: { id: string; componentId: string; nodeLocation: string | null };
+};
+type RegisteredComponentsPayload = { definitions: ComponentDefinition[] };
+type RequestComponentTreeUpdatePayload = undefined;
+type ComponentDragCanceledPayload = undefined;
+type ComponentDroppedPayload = {
+  node: ExperienceTreeNode;
+  index: number;
+  parentNode: {
+    type?: ExperienceTreeNode['type'] | 'root';
+    data: { blockId?: string; id?: string };
+  };
+};
+type ComponentMovedPayload = {
+  nodeId: string;
+  sourceParentId: string;
+  destinationParentId: string;
+  sourceIndex: number;
+  destinationIndex: number;
+};
+type CanvasReloadPayload = undefined;
+type CanvasErrorPayload = Error;
+type UpdateSelectedComponentCoordinatesPayload = {
+  selectedNodeCoordinates: DOMRect;
+  selectedAssemblyChildCoordinates: DOMRect | null;
+  parentCoordinates: DOMRect | null;
+};
+type CanvasScrollPayload = (typeof SCROLL_STATES)[keyof typeof SCROLL_STATES];
+type ComponentMoveStartedPayload = undefined;
+type ComponentMoveEndedPayload = undefined;
+type OutsideCanvasClickPayload = { outsideCanvasClick: boolean };
+type SDKFeaturesPayload = Record<string, unknown>;
+type RequestEntitiesPayload = {
+  entityIds: string[];
+  entityType: 'Entry' | 'Asset';
+  locale: string;
+};
+
+type OUTGOING_EVENT_PAYLOADS = {
+  connected: ConnectedPayload;
+  registerDesignTokens: DesignTokensPayload;
+  registeredBreakpoints: RegisteredBreakpointsPayload;
+  mouseMove: MouseMovePayload;
+  newHoveredElement: NewHoveredElementPayload;
+  componentSelected: ComponentSelectedPayload;
+  registeredComponents: RegisteredComponentsPayload;
+  requestComponentTreeUpdate: RequestComponentTreeUpdatePayload;
+  componentDragCanceled: ComponentDragCanceledPayload;
+  componentDropped: ComponentDroppedPayload;
+  componentMoved: ComponentMovedPayload;
+  canvasReload: CanvasReloadPayload;
+  canvasError: CanvasErrorPayload;
+  updateSelectedComponentCoordinates: UpdateSelectedComponentCoordinatesPayload;
+  canvasScrolling: CanvasScrollPayload;
+  componentMoveStarted: ComponentMoveStartedPayload;
+  componentMoveEnded: ComponentMoveEndedPayload;
+  outsideCanvasClick: OutsideCanvasClickPayload;
+  sdkFeatures: SDKFeaturesPayload;
+  REQUEST_ENTITIES: RequestEntitiesPayload;
+};
+
+export type SendMessageParams = <T extends OutgoingEvent>(
+  eventType: T,
+  data: OUTGOING_EVENT_PAYLOADS[T],
+) => void;
+
+export type OutgoingMessage = {
+  [K in keyof OUTGOING_EVENT_PAYLOADS]: {
+    source: 'customer-app';
+    eventType: K;
+    payload: OUTGOING_EVENT_PAYLOADS[K];
+  };
+}[keyof OUTGOING_EVENT_PAYLOADS];
```

Comment:
Sometimes I just feel like TypeScript was created to confuse people. These kinda hacks look like magic to me 😅